dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp

```diff
@@ -0,0 +1,988 @@
+        animation.Duration(m_playDuration);
+        auto linearEasing = compositor.CreateLinearEasingFunction();
+
+        // Play from fromProgress.
+        animation.InsertKeyFrame(0, m_fromProgress);
+
+        // from > to is treated as playing from fromProgress to the end, then playing from
+        // the beginning to toProgress. Insert extra keyframes to do that.
+        if (m_fromProgress > m_toProgress)
+        {
+            // Play to the end.
+            auto timeToEnd = (1 - m_fromProgress) / ((1 - m_fromProgress) + m_toProgress);
+            animation.InsertKeyFrame(timeToEnd, 1, linearEasing);
+            // Jump to the beginning.
+            animation.InsertKeyFrame(timeToEnd + FLT_EPSILON, 0, linearEasing);
+        }
+
+        // Play to toProgress
+        animation.InsertKeyFrame(1, m_toProgress, linearEasing);
+
+        if (m_looped)
+        {
+            animation.IterationBehavior(winrt::AnimationIterationBehavior::Forever);
+        }
+        else
+        {
+            animation.IterationBehavior(winrt::AnimationIterationBehavior::Count);
+            animation.IterationCount(1);
+        }
+
+        // Create a batch so that we can know when the animation finishes. This only
+        // works for non-looping animations (the batch completes immediately
+        // for looping animations).
+        m_batch = m_looped
+            ? nullptr
+            : compositor.CreateScopedBatch(winrt::CompositionBatchTypes::Animation);
+
+        // Start the animation and get the controller.
+        m_owner.m_progressPropertySet.StartAnimation(L"Progress", animation);
+
+        m_controller = m_owner.m_progressPropertySet.TryGetAnimationController(L"Progress");
+
+        if (m_isPaused || m_isPausedBecauseHidden)
+        {
+            // The play was paused before it was started.
+            m_controller.Pause();
+        }
+
+        // Set the playback rate.
+        auto playbackRate = static_cast<float>(m_owner.PlaybackRate());
+        m_controller.PlaybackRate(playbackRate);
+
+        if (playbackRate < 0)
+        {
+            // Play from end to beginning if playing in reverse.
+            m_controller.Progress(1);
+        }
+
+        if (m_batch)
+        {
+            // Subscribe to the batch completed event.
+            m_batchCompletedToken = m_batch.Completed([this](winrt::IInspectable const&, winrt::CompositionBatchCompletedEventArgs const&)
+            {
+                // Complete the play when the batch completes.
+                // Do not do anything after calling Complete()... the object is destructed already.
+                this->Complete();
+            });
+            // Indicate that nothing else is going into the batch.
+            m_batch.End();
+        }
+
+        m_owner.IsPlaying(true);
+    }
+}
+
+bool AnimatedVisualPlayer::AnimationPlay::IsCurrentPlay()
+{
+    return &*(m_owner.m_nowPlaying) == this;
```

Comment:
The "&*" syntax seems odd. Assuming that shared_ptr doesn't have the comparison overload you're looking for can you use the .get() instead?

```suggestion
    return m_owner.m_nowPlaying.get() == this;
```